src/irmin-pack/unix/control_file.ml

```diff
@@ -299,35 +299,43 @@ module Make (Serde : Serde.S) (Io : Io.S) = struct
   module Io = Io
 
   type payload = Serde.payload
-  type t = { io : Io.t; mutable payload : payload }
+
+  type t = {
+    mutable io : Io.t;
+    mutable payload : payload;
+    path : string;
+    tmp_path : string;
+  }
 
   let write io payload =
     let s = Serde.to_bin_string payload in
-    (* The data must fit inside a single page for atomic updates of the file.
-       This is only true for some file systems. This system will have to be
-       reworked for [V4]. *)
-    assert (String.length s <= Io.page_size);
-
     Io.write_string io ~off:Int63.zero s
 
+  let set_payload t payload =
```

Comment:
I think the simplest solution is to check `if readonly t` and then return the `Ro_not_allowed` error here. Unfortunately, the Io error won't be triggered since you are writing to a new file.